Digital Credentials: implement API using the Identity Document Services framework

marcoscaceres opened the pull request:
#### c4e67707a3a3d9e0550a16d6545bece29304e372
<pre>
Digital Credentials: implement API using the Identity Document Services framework
<a href="https://bugs.webkit.org/show_bug.cgi?id=294237">https://bugs.webkit.org/show_bug.cgi?id=294237</a>
<a href="https://rdar.apple.com/151879131">rdar://151879131</a>

Reviewed by Abrar Rahman Protyasha.

Implements the Digital Credentials API using the new Identity Document Services framework.

On iOS, this allows the user to select digital credentials from application that serve as document providers.
On macOS, credentiala selection cban be performed over CTAP with a compatible device by scanning a QR code.

Canonical link: <a href="https://commits.webkit.org/296741@main">https://commits.webkit.org/296741@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a96a7d69140e7861197e10c9e16ccc7f6b0eb375

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109366 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29024 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19453 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/114571 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59608 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29703 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37612 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83116 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112314 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23648 "Found 1 new test failure: fast/filter-image/clipped-filter.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98506 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63573 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23033 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16649 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/59194 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93016 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16691 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117683 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36407 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92124 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36778 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94768 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91939 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23427 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36868 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14618 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/32216 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36300 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41776 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35973 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39310 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37675 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/UIProcess/DigitalCredentials/WKDigitalCredentialsPicker.h:
```diff
@@ -52,16 +55,14 @@ struct OpenID4VPRequest;
-- (void)presentWithRequestData:(const WebCore::DigitalCredentialsRequestData&)requestData completionHandler:(CompletionHandler<void(Expected<WebCore::DigitalCredentialsResponseData, WebCore::ExceptionData>&&)>&&)completionHandler;
-- (void)dismissWithCompletionHandler:(CompletionHandler<void(bool)>&&)completionHandler;
-
+- (void)presentWithRequestData:(const WebCore::DigitalCredentialsRequestData&)requestData completionHandler:(WTF::CompletionHandler<void(Expected<WebCore::DigitalCredentialsResponseData, WebCore::ExceptionData>&&)>&&)completionHandler;
```
We want to avoid using the `WTF` namespace prefix in our project. We have `using WTF::CompletionHandler;` in `wtf/Forward.h`, so feel free to drop:
```suggestion
- (void)presentWithRequestData:(const WebCore::DigitalCredentialsRequestData&)requestData completionHandler:(CompletionHandler<void(Expected<WebCore::DigitalCredentialsResponseData, WebCore::ExceptionData>&&)>&&)completionHandler;
```

aprotyas commented on Source/WebKit/UIProcess/DigitalCredentials/WKDigitalCredentialsPicker.h:
```diff
@@ -52,16 +55,14 @@ struct OpenID4VPRequest;
-- (void)dismissWithCompletionHandler:(CompletionHandler<void(bool)>&&)completionHandler;
-
+- (void)presentWithRequestData:(const WebCore::DigitalCredentialsRequestData&)requestData completionHandler:(WTF::CompletionHandler<void(Expected<WebCore::DigitalCredentialsResponseData, WebCore::ExceptionData>&&)>&&)completionHandler;
+- (void)dismissWithCompletionHandler:(WTF::CompletionHandler<void(bool)>&&)completionHandler;
```
Same here
```suggestion
- (void)dismissWithCompletionHandler:(CompletionHandler<void(bool)>&&)completionHandler;
```

aprotyas commented on Source/WebKit/UIProcess/DigitalCredentials/WKDigitalCredentialsPicker.mm:
```diff
@@ -158,13 +289,130 @@ - (void)dismissWithCompletionHandler:(CompletionHandler<void(bool)> &&)completio
+        }
+
+        LOG(DigitalCredentials, "The document provider returned response data: %s.", responseData.utf8().data());
+        RetainPtr<NSString> protocol = response.protocolString;
```
You can safely drop the template arg here (CTAD)
```suggestion
        RetainPtr protocol = response.protocolString;
```

aprotyas reviewed (approved):
LGTM with a couple of small comments, hopefully we can get @rr-codes's blessings with the WebKitSwift changes!

marcoscaceres committed:
Digital Credentials: implement API using the Identity Document Services framework
https://bugs.webkit.org/show_bug.cgi?id=294237
rdar://151879131

Reviewed by Abrar Rahman Protyasha.

Implements the Digital Credentials API using the new Identity Document Services framework.

On iOS, this allows the user to select digital credentials from application that serve as document providers.
On macOS, credentiala selection cban be performed over CTAP with a compatible device by scanning a QR code.

Canonical link: https://commits.webkit.org/296741@main
